android_mic_streaming/app/src/main/java/org/mozilla/deepspeechdemo/MainActivity.kt

```diff
@@ -0,0 +1,137 @@
+import android.media.AudioFormat
+import android.media.AudioRecord
+import android.media.MediaRecorder
+import android.os.Build
+import android.os.Bundle
+import android.view.View
+import androidx.appcompat.app.AppCompatActivity
+import androidx.core.app.ActivityCompat
+import kotlinx.android.synthetic.main.activity_main.*
+import org.mozilla.deepspeech.libdeepspeech.DeepSpeechModel
+import org.mozilla.deepspeech.libdeepspeech.DeepSpeechStreamingState
+
+
+class MainActivity : AppCompatActivity() {
+    private var model: DeepSpeechModel? = null
+    private var streamContext: DeepSpeechStreamingState? = null
+
+    private val BEAM_WIDTH = 50
+    private val LM_ALPHA = 0.75f
+    private val LM_BETA = 1.85f
+
+    private val RECORDER_SAMPLERATE = 16000
+    private val RECORDER_CHANNELS: Int = AudioFormat.CHANNEL_IN_MONO
+    private val RECORDER_AUDIO_ENCODING: Int = AudioFormat.ENCODING_PCM_16BIT
+    private var recorder: AudioRecord? = null
+    private var recordingThread: Thread? = null
+    private var isRecording: Boolean = false
+
+    private var NUM_BUFFER_ELEMENTS = 1024
+    private var BYTES_PER_ELEMENT = 2 // 2 bytes (short) because of 16 bit format
+
+    private fun checkPermissions() {
+        val permissions = arrayOf(
+            Manifest.permission.RECORD_AUDIO,
+            Manifest.permission.READ_EXTERNAL_STORAGE
+        )
+
+        // permission is automatically granted on sdk < 23 upon installation
+        if (Build.VERSION.SDK_INT >= 23)
+        {
+            for (permission in permissions) {
+                if (checkSelfPermission(permission) != PackageManager.PERMISSION_GRANTED) {
+                    ActivityCompat.requestPermissions(this, arrayOf(permission), 3)
+                }
+            }
+        }
+    }
+
+    private fun transcribe() {
+        val audioData = ShortArray(NUM_BUFFER_ELEMENTS)
+
+        while (isRecording) {
+            recorder?.read(
+                audioData,
+                0,
+                NUM_BUFFER_ELEMENTS
+            )
+            model?.feedAudioContent(streamContext, audioData, audioData.size)
+            val decoded = model?.intermediateDecode(streamContext)
+            transcription.text = decoded
+        }
+    }
+
+    private fun createModel() {
+        val modelsPath = modelsPathInput.text.toString()
+        model = DeepSpeechModel(modelsPath + "/output_graph.tflite", BEAM_WIDTH)
+        model?.enableDecoderWihLM(modelsPath + "/lm.binary", modelsPath + "/trie", LM_ALPHA, LM_BETA)
+    }
+
+    private fun startListening() {
+        btnStartInference.text = "Stop Recording"
+
+        status.text = "Creating model...\n"
+
+        if (model == null) {
+            createModel()
+            status.append("Created model.\n")
+        } else {
+            status.append("Model already created.\n")
+        }
+
+        streamContext = model?.createStream()
+
+        recorder = AudioRecord(
+            MediaRecorder.AudioSource.MIC,
```

Comment:
You might want `MediaRecorder.AudioSource.VOICE_RECOGNITION`, this has proven to improve results on all tests I did.

Reply:
Okay nice I'll check it out.